File: 238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int l = nums.size();
        vector<int> prods(l);
        for (int i = 0, prod = 1; i < l; i++) {
            prods[i] = prod;
            prod *= nums[i];
        }
        for (int i = l - 1, prod = 1; i >= 0; i--) {
            prods[i] *= prod;
            prod *= nums[i];
        }
        return prods;
    }
};

File: 238. Product of Array Except Self/238. Product of Array Except Self.rb
# @param {Integer[]} nums
# @return {Integer[]}
def product_except_self(nums)
    l = nums.size
    prods = [1] * l
    prodl = prodr = 1
    1.upto(l - 1) { |i|
        prods[i] *= prodl *= nums[i - 1]
        prods[~i] *= prodr *= nums[-i]
    }
    return prods
end
